browser/caliban.py

```diff
@@ -539,210 +417,259 @@ def action_threshold(self, y1, x1, y2, x2, frame, label):
+            self.state.cell_info[self.feature][add_label]['frames'] = updated_frames
         # cell does not exist anywhere in npz:
         except KeyError:
-            self.file.cell_info[self.feature][add_label] = {
+            self.state.cell_info[self.feature][add_label] = {
                 'label': str(add_label),
                 'frames': [frame],
                 'slices': ''
             }
-            self.file.cell_ids[self.feature] = np.append(self.file.cell_ids[self.feature],
-                                                         add_label)
+            self.state.cell_ids[self.feature] = np.append(self.state.cell_ids[self.feature],
+                                                          add_label)
 
         # if adding cell, frames and info have necessarily changed
-        self._y_changed = self.info_changed = True
+        self.y_changed = self.info_changed = True
 
     def del_cell_info(self, del_label, frame):
         """Remove a cell from the npz"""
         # remove cell from frame
-        old_frames = self.file.cell_info[self.feature][del_label]['frames']
+        old_frames = self.state.cell_info[self.feature][del_label]['frames']
         updated_frames = np.delete(old_frames, np.where(old_frames == np.int64(frame))).tolist()
-        self.file.cell_info[self.feature][del_label]['frames'] = updated_frames
+        self.state.cell_info[self.feature][del_label]['frames'] = updated_frames
 
         # if that was the last frame, delete the entry for that cell
-        if self.file.cell_info[self.feature][del_label]['frames'] == []:
-            del self.file.cell_info[self.feature][del_label]
+        if self.state.cell_info[self.feature][del_label]['frames'] == []:
+            del self.state.cell_info[self.feature][del_label]
 
             # also remove from list of cell_ids
-            ids = self.file.cell_ids[self.feature]
-            self.file.cell_ids[self.feature] = np.delete(ids, np.where(ids == np.int64(del_label)))
+            ids = self.state.cell_ids[self.feature]
+            self.state.cell_ids[self.feature] = np.delete(ids,
+                                                          np.where(ids == np.int64(del_label)))
 
         # if deleting cell, frames and info have necessarily changed
-        self._y_changed = self.info_changed = True
+        self.y_changed = self.info_changed = True
 
     def create_cell_info(self, feature):
         """Make or remake the entire cell info dict"""
-        self.file.create_cell_info(feature)
+        self.state.create_cell_info(feature)
         self.info_changed = True
 
 
 class TrackEdit(BaseEdit):
-    def __init__(self, file_, output_bucket):
-        super(TrackEdit, self).__init__(file_, output_bucket)
+    def __init__(self, project):
+        super(TrackEdit, self).__init__(project)
 
-        self.scale_factor = 2
+        # TODO: do we need this?
+        # self.scale_factor = 2
```

Comment:
Looks great! Thanks for implementing the current-frame workaround. Although watershed bugs are still possible, they seem unlikely and pretty minor, so I'm fine with it for the time being. Trk files still need `scale_factor` but once the trk frontend is switched to make use of `CanvasState`, I don't think any of the objects will need to store a `scale_factor` attribute.

Reply:
Just for reference, why does tracking need `scale_factor = 2`?

Reply:
The scaling factor of 2 specifically is historically what was used for trk files. Having a `scale_factor` attribute at all is from pre-`CanvasState` frontend code. The scale factor gets sent in the first trk payload to the frontend as part of initializing the canvas with the correct size. The scale factor is subsequently used as part of image processing (frontend) and mapping the cursor location to the correct point in the image (frontend and some places in backend). The changes that `CanvasState`/zooming bring remove the need for `scale_factor` to be stored in the backend--it becomes implicitly 1, nothing is rescaled. The scale factor in the frontend becomes "whatever makes the image fit". So, tracking only needs `scale_factor = 2` for now until it can be refactored out, which is in progress

Reply:
OK let's uncomment it here and add a TODO comment to remove it during the next refactor.